src/basic/hashmap.c

```diff
@@ -766,6 +777,24 @@ static struct HashmapBase *hashmap_base_new(const struct hash_ops *hash_ops, Ite
         assert_se(pthread_mutex_unlock(&hashmap_debug_list_mutex) == 0);
 #endif
 
+        /* If a cache is desired, we allocate it here as part of hashmap creation
+         * since only a single cache instance per hashmap can really be sanely
+         * supported via the single dirty bit.  The returned cache must be freed
+         * separately from the hashmap it caches. */
+        if (res_cache) {
+                IteratedCache *cache;
+
+                cache = new0(IteratedCache, 1);
+                if (!cache) {
+                        internal_hashmap_free(h);
+                        return NULL;
+                }
+
+                cache->hashmap = h;
+
+                *res_cache = cache;
+        }
```

Comment:
why allocate this here at all? wouldn't it be nicer to just allocate this implicitly when iterated_cache_get() is called for the first time? that way the Hashmap+Set would gain the dirty bit, but would otherwise pretty much remain as they are, and the array cache would be mostly independent, except that it makes use of the dirty bit.

Reply:
The issue with implicitly creating the `IteratedCache` when `iterated_cache_get()` is first called is there's nothing preventing creating multiple of these for a single hashmap without adding a pointer in the base hashmap struct where the single instance `IteratedCache` can be referenced upon implicit creation.

In the spirit of not enlarging the base hashmap structure for this exceptionally utilized feature, it seemed acquiring the `IteratedCache` instance at hashmap create time was an elegant way to prevent programmer error, enforcing a singleton cache per hashmap without adding another pointer to the base hashmap struct.

Also note that if `iterated_cache_get()` needed to create the cache implicitly, it would need to be supplied the hashmap - something currently avoided.  Unless I'm mistaken, the current hashmap implementation would then require a variant of `iterated_cache_get()` for every hashmap variant.

This PR avoids all that by returning a hashmap variant-agnostic `IteratedCache` when the hashmap variant is created, and then we can use the same `iterated_cache_get()` against the cache, regardless of which hashmap variant it's caching.

I'm not however averse to hiding the cache parameter behind macros for all the hashmap variants, and adding `_full()` versions where the cache may be requested.   Though I do believe there is value in having the cache argument visible for discoverability reasons, as a means of getting authors to think about their hashmap usage and if the cache is applicable to what they're doing.  Hiding it behind macros is likely to just leave it largely unused / forgotten outside of the specific use case it's been introduced for.

/cc @poettering

Reply:
sorry for not reviewing this more timely...

Adding the dirty bit would be free actually, HashmapBase has a bunch of bitfields at the end defined currently, and if yum sum them up it's 7 bits. On most archs this would probably be padded to 32bit anyway, hence using a single bit more there is entirely fine...

Reply:
ah, you found that already, ignore me